# Loan Default Risk: Exploratory Data Analysis and Baseline Model

## 1. Project Objective and Research Question

### Research Question

**Can we identify borrowers who are at higher risk of default so that a financial institution can make better lending and credit-risk decisions?**

### Project Objective

The objective of this project is to use historical borrower and loan data to investigate characteristics and patterns associated with loan default. The analysis will use data cleaning, exploratory data analysis, feature engineering, and visualization techniques to develop a better understanding of the factors related to default risk.

A baseline classification model will then be developed to establish an initial benchmark for predicting loan default. The model's performance will be evaluated using an appropriate evaluation metric, with the choice of metric justified according to the classification problem and business context.

The results of the analysis will be used to answer the research question and provide information that may help financial institutions better understand credit risk and make more informed lending decisions.



## 2. Data Used

The analysis uses the **Loan Default Dataset**, publicly available through **Kaggle**. The dataset contains historical information about loan applications, borrowers, financial characteristics, credit information, property characteristics, and loan outcomes.

**Source:** [Kaggle — Loan Default Dataset](https://www.kaggle.com/datasets/yasserh/loan-default-dataset?utm_source=chatgpt.com)

The dataset contains **148,670 observations and 34 features**, including 33 predictor variables and one target variable (`Status`).

The following section describes the features and their meanings. Where the source provides coded categories without defining the exact meaning of each code, the original codes will be retained and treated as categorical variables rather than assigning unsupported interpretations.


| Feature                     | Description                                                                                                                                                                                            |
| --------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| `ID`                        | Unique identifier for each loan application.                                                                                                                                                           |
| `year`                      | Year in which the loan application was recorded.                                                                                                                                                       |
| `loan_limit`                | Indicates whether the loan is conforming (`cf`) or non-conforming (`ncf`).                                                                                                                             |
| `Gender`                    | Gender category of the applicant.                                                                                                                                                                      |
| `approv_in_adv`             | Indicates whether the loan was approved in advance.                                                                                                                                                    |
| `loan_type`                 | Category representing the type of loan. The dataset uses coded categories such as Type1, Type2, and Type3; the source does not provide a reliable explanation of what each individual code represents. |
| `loan_purpose`              | Category representing the purpose of the loan. The dataset uses coded categories such as P1–P4; the source does not provide a reliable explanation of what each individual code represents.            |
| `Credit_Worthiness`         | Category representing the applicant's credit worthiness. The dataset uses coded categories such as L1 and L2.                                                                                          |
| `open_credit`               | Indicates whether the applicant has open credit accounts.                                                                                                                                              |
| `business_or_commercial`    | Indicates whether the loan is classified as business/commercial or personal.                                                                                                                           |
| `loan_amount`               | Amount of money borrowed.                                                                                                                                                                              |
| `rate_of_interest`          | Interest rate associated with the loan.                                                                                                                                                                |
| `Interest_rate_spread`      | Difference between the loan's interest rate and a benchmark rate.                                                                                                                                      |
| `Upfront_charges`           | Charges paid or assessed at the beginning of the loan.                                                                                                                                                 |
| `term`                      | Length of the loan in months.                                                                                                                                                                          |
| `Neg_ammortization`         | Indicates whether the loan permits negative amortization.                                                                                                                                              |
| `interest_only`             | Indicates whether the loan includes an interest-only payment option.                                                                                                                                   |
| `lump_sum_payment`          | Indicates whether a lump-sum payment is part of the loan structure.                                                                                                                                    |
| `property_value`            | Value of the property associated with the loan.                                                                                                                                                        |
| `construction_type`         | Type of property construction, such as site-built or manufactured.                                                                                                                                     |
| `occupancy_type`            | Indicates whether the property is used as a primary residence, secondary residence, or investment property.                                                                                            |
| `Secured_by`                | Type of property used as collateral, such as a home or land.                                                                                                                                           |
| `total_units`               | Number of housing units associated with the property.                                                                                                                                                  |
| `income`                    | Applicant income.                                                                                                                                                                                      |
| `credit_type`               | Type of credit bureau information used for the applicant.                                                                                                                                              |
| `Credit_Score`              | Applicant's credit score.                                                                                                                                                                              |
| `co-applicant_credit_type`  | Type of credit bureau information associated with the co-applicant.                                                                                                                                    |
| `age`                       | Age group of the applicant.                                                                                                                                                                            |
| `submission_of_application` | Indicates the channel through which the application was submitted.                                                                                                                                     |
| `LTV`                       | Loan-to-value ratio, representing the loan amount relative to the property value.                                                                                                                      |
| `Region`                    | Geographic region associated with the loan.                                                                                                                                                            |
| `Security_Type`             | Type of security or collateral arrangement associated with the loan.                                                                                                                                   |
| `Status`                    | Target variable indicating loan outcome: `1` represents default and `0` represents non-default.                                                                                                        |
| `dtir1`                     | Debt-to-income ratio, representing debt obligations relative to income.                                                                                                                                |


Three categorical features—**loan_type**, **loan_purpose**, and **Credit_Worthiness**—contain coded categories whose exact real-world meanings are not fully defined in the available dataset documentation. For example, **loan_type** contains Type1, Type2, and Type3, while **loan_purpose** contains P1–P4.

Rather than assigning meanings that are not supported by the available documentation, this analysis will retain the original category labels and treat these variables as categorical features. The analysis will focus on understanding the relationship between these categories and the target variable, **Status**, including whether different categories are associated with different default rates. These features will therefore be included in the exploratory data analysis, and any findings will be reported using the original category labels.


## 3. Analysis Approach

The analysis will follow a structured process designed to move from understanding the raw data to developing an initial predictive model.

First, the dataset will be inspected and cleaned to address issues such as missing values, duplicate observations, incorrect data types, and potentially invalid values. Next, exploratory data analysis will be conducted to examine distributions, patterns, and relationships among borrower characteristics, loan characteristics, and default outcomes.

Outliers and unusual observations will also be investigated and feature engineering will then be used where appropriate.

Following the exploratory analysis, the prepared data will be used to develop a baseline classification model. The model will be evaluated using an appropriate performance metric, and the result will be interpreted in relation to the research question and business problem.

## 4. Data Cleaning and Preparation

Before conducting the exploratory analysis, the dataset will be assessed and prepared to ensure that the analysis is based on reliable and consistent data. This section will examine missing values, duplicate observations, incorrect data types, inconsistent values, and other potential data-quality issues. Appropriate methods will be used to handle missing information and remove or correct problematic observations when justified.

In [ ]:
# Import the libraries required for data cleaning and analysis.
import pandas as pd
import numpy as np


# Load the Loan Default Dataset.
df = pd.read_csv("data/Loan Default dataset.csv")

# Display the first five observations.
df.head()

In [ ]:
# Display the number of observations and features.
print(f"Number of observations: {df.shape[0]:,}")
print(f"Number of features: {df.shape[1]}")

# Review column names, data types, and non-null counts.
df.info()

In [ ]:
## Analyze Missing Values

# Calculate the number of missing values in each column.
missing_count = df.isnull().sum()

# Calculate the percentage of missing values in each column.
missing_percentage = (missing_count / len(df)) * 100

# Combine the results into a summary table.
missing_summary = pd.DataFrame({"Missing Values": missing_count, "Missing Percentage": missing_percentage})

# Display only features that contain missing values,
# ordered from highest to lowest percentage.
missing_summary = missing_summary[ missing_summary["Missing Values"] > 0].sort_values("Missing Percentage", ascending=False)

missing_summary

<span style="color:cadetblue;"><b></b>
The largest missingness is in **Upfront_charges**, **Interest_rate_spread**, and **rate_of_interest**, while several other variables have very little missing data. We need to understand the values and patterns before choosing a cleaning method.</span>

In [120]:
# Check for Duplicate Observations

# Count completely duplicated rows.
duplicate_count = df.duplicated().sum()

print(f"Number of duplicate observations: {duplicate_count:,}")

Number of duplicate observations: 0


In [121]:
# ## Check Target Class Distribution

# Calculate the count and percentage of each target class.
status_distribution = pd.DataFrame({
    "Count": df["Status"].value_counts().sort_index(),
    "Percentage": df["Status"].value_counts(normalize=True).sort_index() * 100})

status_distribution

,Count,Percentage
Status,,
0,112031,75.355485
1,36639,24.644515


<span style="color:cadetblue;"><b></b>
The target variable is imbalanced, which is important for the modeling stage because accuracy alone may not adequately evaluate model performance. This distribution will be considered when selecting and justifying the appropriate evaluation metric.</span>

### Examine Categorical features

In [ ]:
# Examine Categorical features

# Identify categorical features.
categorical_columns = df.select_dtypes(include="object").columns

# Calculate and display the distribution for each categorical feature.
for column in categorical_columns:
    value_counts = df[column].value_counts(dropna=False)
    value_percentages = (df[column].value_counts(dropna=False, normalize=True) * 100 )

    # Count Status 0 and Status 1 within each category.
    status_counts = (df.groupby([column, "Status"], dropna=False) .size().unstack(fill_value=0) )

    # Make sure both Status 0 and Status 1 columns exist.
    if 0 not in status_counts.columns:
        status_counts[0] = 0

    if 1 not in status_counts.columns:
        status_counts[1] = 0

    # Calculate Status percentages within each category.
    status_percentages = status_counts.div(
        status_counts.sum(axis=1), axis=0) * 100

    category_summary = pd.DataFrame({
        "Count": value_counts,
        "Percentage": value_percentages.round(2),
        "Status 0 Count": status_counts[0],
        "Status 0 %": status_percentages[0].round(2),
        "Status 1 Count": status_counts[1],
        "Status 1 %": status_percentages[1].round(2)})

    unique_values = df[column].nunique(dropna=False)

    print(f"\n{column} — Unique Values: {unique_values}")
    display(category_summary)

<span style="color:cadetblue;"><b></b>
We noticed that **construction_type**, **Secured_by**, and **Security_Type** each contain only 33 observations in one category, while almost all remaining observations fall into the other category.
Because these features are highly concentrated in one category, they were investigated further to determine whether they provide meaningful information or should be removed.</span>


In [ ]:
# Inspect the 33 observations in the rare categories

rare_33 = df[
    (df["construction_type"] == "mh") |
    (df["Secured_by"] == "land") |
    (df["Security_Type"] == "Indriect")
]

display(
    rare_33[
        ["ID", "construction_type", "Secured_by", "Security_Type", "Status"]
    ]
)

In [ ]:
# Investigate the 33 rare observations and their business characteristics

# Identify the three rare categories.
mh_ids = set(df.loc[df["construction_type"] == "mh", "ID"])
land_ids = set(df.loc[df["Secured_by"] == "land", "ID"])
indirect_ids = set(df.loc[df["Security_Type"] == "Indriect", "ID"])

# Check whether the three categories contain the same observations.
common_ids = mh_ids & land_ids & indirect_ids

print("Construction type = mh:", len(mh_ids))
print("Secured by = land:", len(land_ids))
print("Security type = Indriect:", len(indirect_ids))
print("Common observations across all three:", len(common_ids))

# Display the complete rows for the common observations.
rare_33 = df[df["ID"].isin(common_ids)].copy()

print("\nThe 33 common observations:")
display(rare_33)

# Compare their Status distribution.
print("\nStatus distribution:")
display(rare_33["Status"].value_counts().sort_index())

# Display key business-related variables for these 33 observations.
business_features = [
    "ID",
    "loan_amount",
    "loan_purpose",
    "occupancy_type",
    "property_value",
    "income",
    "Credit_Score",
    "LTV",
    "dtir1",
    "credit_type",
    "Gender",
    "age",
    "construction_type",
    "Secured_by",
    "Security_Type",
    "Status"
]

print("\nKey business characteristics of the 33 observations:")
display(rare_33[business_features])

<span style="color:cadetblue;"><b></b>
The 33 observations share the same rare categories across **construction_type**, **Secured_by**, and **Security_Type**, suggesting they may represent a specific loan/property subgroup.These observations appear valid and consistent, so no cleaning or removal is required. 
However, from a feature perspective, each feature has only 33 observations in one category, while nearly all remaining observations belong to the other category, so these features provide very limited variation and were removed from the baseline dataset.Therefore, these features were treated as representing essentially one category and removed from the baseline dataset.</span>

In [125]:
# Remove the three highly redundant features

df = df.drop(columns=["construction_type", "Secured_by", "Security_Type"]).copy()

# Confirm the number of remaining features.
print(f"Number of features after removal: {df.shape[1]}")

Number of features after removal: 31


### Investigating Numerical Variables

In [ ]:
# Check Numerical Features by Loan Default Status

# Identify numerical features, excluding the target variable.
numerical_columns = df.select_dtypes(
    include=["int64", "float64"]
).columns
numerical_columns = numerical_columns[numerical_columns != "Status"]

results = []

for column in numerical_columns:
    status_0_count = df.loc[df["Status"] == 0, column].notna().sum()
    status_1_count = df.loc[df["Status"] == 1, column].notna().sum()

    status_0_total = (df["Status"] == 0).sum()
    status_1_total = (df["Status"] == 1).sum()

    results.append({
        "Feature": column,
        "Status 0 Count": status_0_count,
        "Status 0 %": round(status_0_count / status_0_total * 100, 2),
        "Status 1 Count": status_1_count,
        "Status 1 %": round(status_1_count / status_1_total * 100, 2)
    })

numerical_status_summary = pd.DataFrame(results)

display(numerical_status_summary)

<span style="color:cadetblue;"><b></b>
**Interest_rate_spread** has no observed values for Status 1; all Status 1 observations are missing for this feature, while its available values correspond only to Status 0. Therefore, it provides no usable numerical information for comparing the two target classes.</span>


In [ ]:
# Check Numerical Variables

# Identify numerical columns.
numerical_columns = df.select_dtypes(include=["int64", "float64"]).columns

# Display descriptive statistics for numerical features.
numerical_summary = df[numerical_columns].describe().T

numerical_summary

<span style="color:cadetblue;"><b></b>
The descriptive statistics identified year as a constant feature and LTV as having an unusually high maximum value. The following check examines their actual values before making any cleaning decisions.</span>

In [ ]:
# ## Inspect Year and LTV Values

# Display the unique value(s) recorded for year.
print("Unique values in year:")
print(df["year"].unique())

# Display selected descriptive information for LTV.
print("\nUnique LTV values:")
print(df["LTV"].nunique())

print("\nSmallest LTV values:")
print(df["LTV"].nsmallest(10).to_list())

print("\nLargest LTV values:")
print(df["LTV"].nlargest(10).to_list())

<span style="color:cadetblue;"><b></b>
**Remove Unnecessary Features**
Before handling missing values, three features are removed from the dataset because they do not provide useful information for the analysis or modeling. The **ID** feature is an identifier assigned to each observation and does not represent a borrower or loan characteristic. The **year** feature contains only one value, 2019, for all observations and therefore has no variation.  Also, **Interest_rate_spread** has no observed values for Status 1. All three features are excluded from the cleaned dataset while the remaining features are retained for further analysis. </span>

In [129]:
# Remove Unnecessary Features

# Remove the identifier and constant year feature.
df = df.drop(columns=["ID", "year", "Interest_rate_spread"]).copy()

# Confirm the remaining number of features.
print(f"Number of features after removal: {df.shape[1]}")

Number of features after removal: 28


### Investigate Potential Outliers

In [ ]:
# Investigate Potential Outliers

# Identify numerical predictor variables.
numerical_predictors = df.select_dtypes(include=["int64", "float64"]).columns

# Exclude the target variable from the outlier analysis.
numerical_predictors = numerical_predictors[numerical_predictors != "Status"]

# Create a list to store the outlier results.
outlier_results = []

# Calculate IQR-based outlier thresholds for each numerical predictor.
for column in numerical_predictors:
    q1 = df[column].quantile(0.25)
    q3 = df[column].quantile(0.75)
    iqr = q3 - q1

    lower_bound = q1 - 1.5 * iqr
    upper_bound = q3 + 1.5 * iqr

    outlier_count = (
        (df[column] < lower_bound) |
        (df[column] > upper_bound)
    ).sum()

    outlier_percentage = (outlier_count / len(df)) * 100

    outlier_results.append({
        "Feature": column,
        "Q1": q1,
        "Q3": q3,
        "IQR": iqr,
        "Lower Bound": lower_bound,
        "Upper Bound": upper_bound,
        "Outlier Count": outlier_count,
        "Outlier Percentage": outlier_percentage
    })

# Create a summary table.
outlier_summary = pd.DataFrame(outlier_results)

# Display the features with the largest number of potential outliers.
outlier_summary.sort_values("Outlier Count",ascending=False)

<span style="color:cadetblue;"><b></b>
The **term** feature shows a high outlier percentage (18.12%) because Q1 and Q3 are both 360, resulting in an IQR of zero. This indicates that most loans have a 360-month term, so we will examine its unique values to determine whether the flagged values are legitimate.</span>


In [ ]:
# Examine Loan Feature

# Count the unique loan-term values.
unique_term_count = df["term"].nunique(dropna=True)

print(f"Number of unique loan-term values: {unique_term_count}")

# Display the frequency of each loan-term value.
term_distribution = ( df["term"].value_counts(dropna=False).sort_index())

term_distribution

<span style="color:cadetblue;"><b></b>
The unique values show that the different loan terms are valid durations, with 360 months being the most common. Therefore, no cleaning or removal is needed for **term**.</span>

<span style="color:cadetblue;"><b></b>
The IQR analysis identified an upper bound of **124.75%** for LTV. Values above this threshold are statistical outliers and also represent loan amounts substantially higher than the recorded property value from a business perspective. Therefore, **LTV values above 125% will be treated as anomalous and removed**, while values at or below the threshold will be retained.</span>


In [132]:
# ## Remove LTV Outliers

# Remove LTV values above 125%.
df = df[df["LTV"] <= 125].copy()

print(f"Remaining observations: {len(df):,}")

Remaining observations: 133,452


### Investigating Missing Values

In [ ]:
# ## Recheck Missing Values After Outlier Removal


# Calculate the number of missing values in each column.
missing_count = df.isnull().sum()

# Calculate the percentage of missing values in each column.
missing_percentage = (missing_count / len(df)) * 100

# Create a summary table.
missing_summary_df = pd.DataFrame({
    "Missing Values": missing_count,
    "Missing Percentage": missing_percentage
})

# Display only features that contain missing values.
missing_summary_df = missing_summary_df[
    missing_summary_df["Missing Values"] > 0
].sort_values("Missing Percentage", ascending=False)

missing_summary_df

In [134]:
# Compare Status distribution before and after removing rows with missing values.

# Status counts before dropping missing values.
before_counts = df["Status"].value_counts().sort_index()

# Drop all rows containing missing values.
df_dropna = df.dropna().copy()

# Status counts after dropping missing values.
after_counts = df_dropna["Status"].value_counts().sort_index()

print("Before dropna:")
print(f"Status 0 (No Default): {before_counts.get(0, 0):,}")
print(f"Status 1 (Default): {before_counts.get(1, 0):,}")
print(f"Total observations: {len(df):,}")

print("\nAfter dropna:")
print(f"Status 0 (No Default): {after_counts.get(0, 0):,}")
print(f"Status 1 (Default): {after_counts.get(1, 0):,}")
print(f"Total observations: {len(df_dropna):,}")

Before dropna:
Status 0 (No Default): 112,022
Status 1 (Default): 21,430
Total observations: 133,452

After dropna:
Status 0 (No Default): 98,186
Status 1 (Default): 0
Total observations: 98,186


<span style="color:cadetblue;"><b></b>
Dropping all rows with missing values removes all 21,430 default observations (Status 1), leaving only Status 0 observations. Therefore, complete-case deletion is not appropriate because it eliminates the target class needed for classification.</span>


In [ ]:
# Investigate whether each numerical feature has usable values
# for both Status 0 and Status 1

numerical_features = df.select_dtypes(
    include=["int64", "float64"]
).columns

numerical_features = numerical_features[
    numerical_features != "Status"
]

results = []

for column in numerical_features:
    
    status_0 = df[df["Status"] == 0][column]
    status_1 = df[df["Status"] == 1][column]
    
    results.append({
        "Feature": column,
        "Status 0 Non-Missing": status_0.notna().sum(),
        "Status 1 Non-Missing": status_1.notna().sum(),
        "Status 0 Missing": status_0.isna().sum(),
        "Status 1 Missing": status_1.isna().sum()
    })

numerical_investigation = pd.DataFrame(results)

display(numerical_investigation)

<span style="color:cadetblue;"><b></b>
The **rate_of_interest** and **Upfront_charges** features have missing values concentrated almost entirely among Status 1 observations. Removing these missing observations would eliminate the default class, so these features were removed instead to preserve both target categories.</span>


In [136]:
# Remove features with missing values heavily concentrated in Status 1.
df = df.drop(columns=["rate_of_interest", "Upfront_charges"]).copy()

# Check the dataset dimensions after removing the two features.
print(f"Number of observations: {df.shape[0]:,}")
print(f"Number of features: {df.shape[1]}")


Number of observations: 133,452
Number of features: 26


In [137]:
# Drop all remaining rows with missing values.

df = df.dropna().copy()

print(f"Observations: {df.shape[0]:,}")
print(f"Features: {df.shape[1]:,}")

print("\nStatus counts:")
print(df["Status"].value_counts().sort_index())

Observations: 120,422
Features: 26

Status counts:
Status
0    100878
1     19544
Name: count, dtype: int64


In [ ]:
# Complete investigation of df

print("Shape:", df.shape)
print("Number of observations:", df.shape[0])
print("Number of features:", df.shape[1])

print("\nData types:")
display(df.dtypes)

print("\nMissing values:")
missing_summary = (
    df.isna()
    .sum()
    .to_frame("Missing Count")
)

missing_summary["Missing %"] = (
    missing_summary["Missing Count"] / len(df) * 100
).round(2)

display(missing_summary[missing_summary["Missing Count"] > 0])

print("\nDuplicate rows:", df.duplicated().sum())

print("\nTarget distribution:")
target_summary = df["Status"].value_counts().to_frame("Count")
target_summary["Percentage"] = (
    target_summary["Count"] / len(df) * 100
).round(2)

display(target_summary)

print("\nNumerical summary:")
display(df.describe().T)

print("\nCategorical summary:")
display(df.select_dtypes(include="object").describe().T)

In [ ]:
# Convert term from float to integer
df["term"] = df["term"].astype(int)

# Dataset size
print("Shape:", df.shape)
print("Number of observations:", df.shape[0])
print("Number of features:", df.shape[1])

# Data types
print("\nData types:")
display(df.dtypes)

# Check missing values for every feature
missing_values = (df.isna().sum().to_frame("Missing Count"))

display(missing_values)

# Duplicate rows
print("\nDuplicate rows:", df.duplicated().sum())

# Target distribution
print("\nTarget distribution:")
target_distribution = (
    df["Status"]
    .value_counts()
    .sort_index()
    .to_frame("Count")
)

display(target_distribution)

# Numerical summary
print("\nNumerical features:")
display(df.describe())

# Categorical summary
print("\nCategorical features:")
display(df.describe(include="object"))

## 5. Exploratory Data Analysis

Exploratory data analysis will be used to understand the characteristics of the dataset and identify patterns and relationships that may help answer the research question.

### Categorical Features and the Target Variable
This analysis examines the relationship between categorical borrower and loan features and the target variable, **Status**. The goal is to identify categories associated with higher or lower observed default rates and determine the size of the differences between categories.


In [ ]:
# Count and percentage of loan default status within each category.

categorical_features = df.select_dtypes(include="object").columns

results = []

for feature in categorical_features:

    summary = (
        df.groupby([feature, "Status"])
        .size()
        .unstack()
    )

    summary["Total"] = summary.sum(axis=1)

    summary["Status 0 %"] = summary[0] / summary["Total"] * 100
    summary["Status 1 %"] = summary[1] / summary["Total"] * 100

    summary = summary.reset_index()

    summary = summary.rename(columns={
        feature: "Category",
        0: "Status 0 Count",
        1: "Status 1 Count"
    })

    summary["Feature"] = feature

    results.append(summary)

categorical_status_summary = pd.concat(results, ignore_index=True)

categorical_status_summary = categorical_status_summary[
    [
        "Feature",
        "Category",
        "Total",
        "Status 0 Count",
        "Status 0 %",
        "Status 1 Count",
        "Status 1 %"
    ]
]

categorical_status_summary["Status 0 %"] = (
    categorical_status_summary["Status 0 %"].round(2)
)

categorical_status_summary["Status 1 %"] = (
    categorical_status_summary["Status 1 %"].round(2)
)

display(categorical_status_summary)

<span style="color:cadetblue;"><b></b>
To identify categorical features that may be associated with loan default, we compared the observed percentage of Status 1 (default) across the categories of each feature. We focused on features where the difference between the highest and lowest observed default rate was greater than approximately 10 percentage points. Based on this comparison, the five features selected for further analysis were **lump_sum_payment**, **Neg_ammortization**, **loan_type**, **business_or_commercial**, and **loan_purpose**. These features were then examined in greater detail through additional visual analysis to better understand their relationship with loan default and their potential usefulness for identifying higher-risk borrowers.</span>


In [ ]:
# Plot Status 0 % and Status 1 % for each category
# within each selected categorical feature.

import plotly.express as px



selected_features = [
    "lump_sum_payment",
    "Neg_ammortization",
    "loan_type",
    "business_or_commercial",
    "loan_purpose"
]

for feature in selected_features:

    plot_data = categorical_status_summary[
        categorical_status_summary["Feature"] == feature
    ].copy()

    plot_data = plot_data.melt(
        id_vars=["Category"],
        value_vars=["Status 0 %", "Status 1 %"],
        var_name="Status",
        value_name="Percentage"
    )

    fig = px.bar(
        plot_data,
        x="Category",
        y="Percentage",
        color="Status",
        barmode="group",
        text="Percentage",
        title=f"{feature}: Loan Default Status by Category",
        labels={
            "Category": "Category",
            "Percentage": "Percentage (%)",
            "Status": "Loan Status"
        },
        height=500
    )

    fig.update_traces(
        texttemplate="%{text:.1f}%",
        textposition="outside"
    )

    fig.update_yaxes(
        range=[0, 100],
        ticksuffix="%"
    )

    fig.update_layout(
        legend_title_text="Loan Status",
        margin=dict(
            l=70,
            r=50,
            t=90,
            b=80
        )
    )

    fig.show()

<span style="color:cadetblue;"><b></b>
For these five features, we observed a noticeable difference in the proportion of Status 0 (no default) and Status 1 (default) across their categories. This means that borrowers in different categories of these features show different observed default rates. From a business perspective, these features may help distinguish groups of borrowers with different levels of observed loan-default risk. These differences suggest that **lump_sum_payment**, **Neg_ammortization**, **loan_type**, **business_or_commercial**, and **loan_purpose** may be useful features for understanding and predicting loan default.</span>

### Numerical Features and the Target Variable

In [ ]:
# Compare numerical features between Status 0 and Status 1.
# These statistics help identify noticeable differences between the two groups.

numerical_features = df.select_dtypes(
    include=["int64", "float64"]).columns

numerical_features = numerical_features[
    numerical_features != "Status"]

results = []

for feature in numerical_features:

    status_0 = df[df["Status"] == 0][feature]
    status_1 = df[df["Status"] == 1][feature]

    results.append({
        "Feature": feature,

        "Status 0 Min": status_0.min(),
        "Status 1 Min": status_1.min(),

        "Status 0 Max": status_0.max(),
        "Status 1 Max": status_1.max(),

        "Status 0 Mean": status_0.mean(),
        "Status 1 Mean": status_1.mean(),

        "Status 0 Median": status_0.median(),
        "Status 1 Median": status_1.median(),

        "Status 0 Std": status_0.std(),
        "Status 1 Std": status_1.std()
    })

numerical_status_summary = pd.DataFrame(results)

# Calculate the percentage difference in the mean.
# Status 0 is used as the reference group.
numerical_status_summary["Percentage Difference vs Status 0"] = (
    (
        numerical_status_summary["Status 1 Mean"]
        - numerical_status_summary["Status 0 Mean"]
    )
    / numerical_status_summary["Status 0 Mean"]
    * 100
)

# Round the numerical results for readability.
numerical_status_summary = numerical_status_summary.round(2)

display(numerical_status_summary)

<span style="color:cadetblue;"><b></b>
To better compare numerical features between defaulted and non-defaulted borrowers, we calculated the percentage difference between their average values. We used Status 0 (non-default) as the reference group to show how much higher or lower the average value is for Status 1 (default). This percentage helps us understand the size and direction of the observed difference between the two groups, regardless of the original unit of each feature. The results show that the average **income** of defaulted borrowers is 16.06% lower than that of non-defaulted borrowers, suggesting that higher income is associated with a lower observed rate of loan default. The average **property value** of defaulted borrowers is 8.58% lower than that of non-defaulted borrowers, suggesting that higher property value is associated with a lower observed rate of loan default. In contrast, the average **dtir1** of defaulted borrowers is 6.07% higher than that of non-defaulted borrowers, suggesting that higher **dtir1** is associated with a higher observed rate of loan default. The average **loan_amount** of defaulted borrowers is 5.69% lower than that of non-defaulted borrowers, suggesting that higher loan amounts are associated with a lower observed rate of loan default. These results highlight numerical features that show noticeable differences between defaulted and non-defaulted borrowers.</span>


In [ ]:
# Plot box plots for income and property_value.
# The y-axis is limited to the 1st–99th percentile range
# so the main distribution is easier to see.

import plotly.express as px

selected_features = [
    "income",
    "property_value",
    "dtir1",
    "loan_amount"
]

for feature in selected_features:

    plot_data = df[["Status", feature]].dropna().copy()

    plot_data["Loan Status"] = plot_data["Status"].map({
        0: "Status 0 (No Default)",
        1: "Status 1 (Default)"
    })

    fig = px.box(
        plot_data,
        x="Loan Status",
        y=feature,
        color="Loan Status",
        points=False,
        title=f"{feature}: Distribution by Loan Status",
        labels={
            "Loan Status": "Loan Status",
            feature: feature
        },
        height=500
    )

    # Zoom the y-axis to the main distribution.
    lower = plot_data[feature].quantile(0.01)
    upper = plot_data[feature].quantile(0.99)

    fig.update_yaxes(
        range=[lower, upper]
    )

    fig.update_layout(
        margin=dict(
            l=70,
            r=50,
            t=90,
            b=80
        )
    )

    fig.show()

In [ ]:
# ## Correlation Matrix of Numerical Features
#
# Visualize the correlations between all numerical features.

numerical_features = df.select_dtypes(include=["int64", "float64"]).columns

correlation_matrix = df[numerical_features].corr()

fig = px.imshow(
    correlation_matrix,
    text_auto=".2f",
    aspect="auto",
    title="Correlation Matrix of Numerical Features",
    labels={"color": "Correlation"}
)

fig.show()

In [ ]:
# ## Correlation Matrix for Numerical Features
#
# Check relationships between all numerical features to identify
# potentially redundant variables before handling missing values.

numerical_features = df.select_dtypes(include=["int64", "float64"]).columns

correlation_matrix = df[numerical_features].corr()

correlation_matrix

<span style="color:cadetblue;"><b></b>
The correlation matrix helps identify relationships between numerical features and their relationship with the target variable, **Status**.
The strongest relationship is between **loan_amount** and **property_value** (0.73), showing that higher property values are generally associated with higher loan amounts.
**property_value** and **LTV** have a negative relationship (-0.44), while **loan_amount** and **income** have a positive relationship (0.45).
These relationships help us understand how borrower and loan characteristics are connected and identify features that may be useful for further analysis and modeling.</span>


### Categorical Features and Numerical Features

In [ ]:
# Compare observed default rates across loan amount ranges and loan types.
# Point size represents average borrower income.

import pandas as pd
import plotly.express as px

plot_data = df[
    ["loan_amount", "income", "loan_type", "Status"]
].dropna().copy()

# Divide loan amounts into 10 ranges.
plot_data["Loan Amount Range"] = pd.qcut(
    plot_data["loan_amount"],
    q=300,
    duplicates="drop"
)

# Calculate default rate and average income for each loan amount range and loan type.
summary = (
    plot_data
    .groupby(["Loan Amount Range", "loan_type"], observed=True)
    .agg(
        Default_Rate=("Status", "mean"),
        Average_Income=("income", "mean")
    )
    .reset_index()
)

summary["Status 1 %"] = summary["Default_Rate"] * 100

# Use the midpoint of each loan amount range for the numerical X-axis.
summary["Loan Amount"] = summary["Loan Amount Range"].apply(
    lambda x: x.mid
)

fig = px.scatter(
    summary,
    x="Loan Amount",
    y="Status 1 %",
    size="Average_Income",
    color="loan_type",
    title="Observed Default Rate by Loan Amount and Loan Type, with Income by Size",
    labels={
        "Loan Amount": "Loan Amount",
        "Status 1 %": "Default Rate (%)",
        "Average_Income": "Average Income",
        "loan_type": "Loan Type"
    },
    opacity=0.75,
    width=900,
    height=500
)

fig.update_yaxes(
    range=[0, 100],
    ticksuffix="%"
)
fig.update_xaxes(
    range=[0, 1000000],
    title="Loan Amount"
)

fig.show()

<span style="color:cadetblue;"><b></b>
Loan amount shows no clear relationship with observed default rate. In contrast, loan type shows a clearer pattern, with Type 1 generally having lower default rates and Type 2 higher default rates.</span>


In [ ]:
# Compare observed default rates across loan amount ranges and lump sum payment.
# Point size represents average borrower income.

plot_data = df[
    ["loan_amount", "income", "lump_sum_payment", "Status"]
].dropna().copy()

plot_data["Loan Amount Range"] = pd.qcut(
    plot_data["loan_amount"],
    q=300,
    duplicates="drop"
)

summary = (
    plot_data
    .groupby(["Loan Amount Range", "lump_sum_payment"], observed=True)
    .agg(
        Default_Rate=("Status", "mean"),
        Average_Income=("income", "mean")
    )
    .reset_index()
)

summary["Status 1 %"] = summary["Default_Rate"] * 100

summary["Loan Amount"] = summary["Loan Amount Range"].apply(
    lambda x: x.mid
)

fig = px.scatter(
    summary,
    x="Loan Amount",
    y="Status 1 %",
    size="Average_Income",
    color="lump_sum_payment",
    title="Observed Default Rate by Loan Amount and Lump Sum Payment, with Income by Size",
    labels={
        "Loan Amount": "Loan Amount",
        "Status 1 %": "Default Rate (%)",
        "Average_Income": "Average Income",
        "lump_sum_payment": "Lump Sum Payment"
    },
    opacity=0.75,
    width=900,
    height=500
)

fig.update_xaxes(
    range=[0, 1000000],
    title="Loan Amount"
)

fig.update_yaxes(
    range=[0, 100],
    ticksuffix="%"
)

fig.show()

<span style="color:cadetblue;"><b></b>
Loan amount does not show a clear relationship with observed default rate. However, lump-sum payment status shows a much stronger association with default. Borrowers in the **lpsm** category generally have substantially higher and more variable default rates. For loan amounts below $0.2 million, many **lpsm** observations show default rates between approximately 60% and 80%, while at higher loan amounts the default rates become more spread out and can reach 100%. In comparison, borrowers in the **not_lpsm** category have default rates that are mostly below 20%. This pattern suggests that lump-sum payment status may be an important feature for identifying borrowers associated with higher default risk.</span>


In [ ]:
# Compare observed default rates across loan amount ranges and negative amortization.
# Point size represents average borrower income.

plot_data = df[
    ["loan_amount", "income", "Neg_ammortization", "Status"]
].dropna().copy()

plot_data["Loan Amount Range"] = pd.qcut(
    plot_data["loan_amount"],
    q=300,
    duplicates="drop"
)

summary = (
    plot_data
    .groupby(["Loan Amount Range", "Neg_ammortization"], observed=True)
    .agg(
        Default_Rate=("Status", "mean"),
        Average_Income=("income", "mean")
    )
    .reset_index()
)

summary["Status 1 %"] = summary["Default_Rate"] * 100

summary["Loan Amount"] = summary["Loan Amount Range"].apply(
    lambda x: x.mid
)

fig = px.scatter(
    summary,
    x="Loan Amount",
    y="Status 1 %",
    size="Average_Income",
    color="Neg_ammortization",
    title="Observed Default Rate by Loan Amount and Negative Amortization, with Income by Size",
    labels={
        "Loan Amount": "Loan Amount",
        "Status 1 %": "Default Rate (%)",
        "Average_Income": "Average Income",
        "Neg_ammortization": "Negative Amortization"
    },
    opacity=0.75,
    width=900,
    height=500
)

fig.update_xaxes(
    range=[0, 1000000],
    title="Loan Amount"
)

fig.update_yaxes(
    range=[0, 100],
    ticksuffix="%"
)

fig.show()

<span style="color:cadetblue;"><b></b>
Loan amount does not show a clear relationship with observed default rate. However, negative amortization status shows a stronger association with default. Borrowers in the **neg_amm** category consistently show higher default rates than those in the **not_neg** category across different loan amounts. The **not_neg** category remains mostly below 20%, while **neg_amm** generally remains above 20% and reaches approximately 60% at some loan amounts. This pattern suggests that negative amortization status may be an important feature for identifying borrowers associated with higher default risk.</span>


In [ ]:
# Compare observed default rates across DTIR1 ranges and loan type.
# Point size represents average borrower income.

plot_data = df[
    ["dtir1", "income", "loan_type", "Status"]
].dropna().copy()

plot_data["DTIR1 Range"] = pd.qcut(
    plot_data["dtir1"],
    q=300,
    duplicates="drop"
)

summary = (
    plot_data
    .groupby(["DTIR1 Range", "loan_type"], observed=True)
    .agg(
        Default_Rate=("Status", "mean"),
        Average_Income=("income", "mean")
    )
    .reset_index()
)

summary["Status 1 %"] = summary["Default_Rate"] * 100

summary["DTIR1"] = summary["DTIR1 Range"].apply(
    lambda x: x.mid
)

fig = px.scatter(
    summary,
    x="DTIR1",
    y="Status 1 %",
    size="Average_Income",
    color="loan_type",
    title="Observed Default Rate by DTIR1 and Loan Type, with Income by Size",
    labels={
        "DTIR1": "DTIR1",
        "Status 1 %": "Default Rate (%)",
        "Average_Income": "Average Income",
        "loan_type": "Loan Type"
    },
    opacity=0.75,
    width=900,
    height=500
)

fig.update_xaxes(
    range=[0, 100],
    title="DTIR1"
)

fig.update_yaxes(
    range=[0, 100],
    ticksuffix="%"
)

fig.show()

<span style="color:cadetblue;"><b></b>
The observed default rate differs across loan types at different DTI levels. Below a DTI of about 50, Type 1 generally shows lower observed default rates than Type 2 and Type 3. Around a DTI of 50–60, this pattern changes, with Type 1 showing substantially higher observed default rates in some groups. Overall, the exploratory analysis suggests that loan type and DTI jointly show different patterns of observed default rates, indicating that their combination may be associated with loan default.</span>


In [ ]:
# Compare observed default rates across DTIR1 ranges and lump sum payment.
# Point size represents average borrower income.

plot_data = df[
    ["dtir1", "income", "lump_sum_payment", "Status"]
].dropna().copy()

plot_data["DTIR1 Range"] = pd.qcut(
    plot_data["dtir1"],
    q=300,
    duplicates="drop"
)

summary = (
    plot_data
    .groupby(["DTIR1 Range", "lump_sum_payment"], observed=True)
    .agg(
        Default_Rate=("Status", "mean"),
        Average_Income=("income", "mean")
    )
    .reset_index()
)

summary["Status 1 %"] = summary["Default_Rate"] * 100

summary["DTIR1"] = summary["DTIR1 Range"].apply(
    lambda x: x.mid
)

fig = px.scatter(
    summary,
    x="DTIR1",
    y="Status 1 %",
    size="Average_Income",
    color="lump_sum_payment",
    title="Observed Default Rate by DTIR1 and Lump Sum Payment, with Income by Size",
    labels={
        "DTIR1": "DTIR1",
        "Status 1 %": "Default Rate (%)",
        "Average_Income": "Average Income",
        "lump_sum_payment": "Lump Sum Payment"
    },
    opacity=0.75,
    width=900,
    height=500
)

fig.update_xaxes(
    range=[0, 100],
    title="DTIR1"
)

fig.update_yaxes(
    range=[0, 100],
    ticksuffix="%"
)

fig.show()

<span style="color:cadetblue;"><b></b>
The observed default rate is consistently higher for borrowers in the **lpsm** category than for those in the **not_lpsm** category across the DTI range from approximately 0 to 60. Within the **lpsm** category, default rates are relatively high at lower DTI levels, decrease across the middle range of approximately 20 to 50, and increase again at higher DTI levels around 50 to 60. The **not_lpsm** category follows a similar pattern, with lower default rates overall but a decrease in the middle DTI range and higher rates toward the upper end. Overall, the pattern suggests that lump-sum payment status is strongly associated with observed default rates, while the default-rate pattern varies across different DTI levels.</span>


In [ ]:
# Compare observed default rates across DTIR1 ranges and negative amortization.
# Point size represents average borrower income.

plot_data = df[
    ["dtir1", "income", "Neg_ammortization", "Status"]
].dropna().copy()

plot_data["DTIR1 Range"] = pd.qcut(
    plot_data["dtir1"],
    q=300,
    duplicates="drop"
)

summary = (
    plot_data
    .groupby(["DTIR1 Range", "Neg_ammortization"], observed=True)
    .agg(
        Default_Rate=("Status", "mean"),
        Average_Income=("income", "mean")
    )
    .reset_index()
)

summary["Status 1 %"] = summary["Default_Rate"] * 100

summary["DTIR1"] = summary["DTIR1 Range"].apply(
    lambda x: x.mid
)

fig = px.scatter(
    summary,
    x="DTIR1",
    y="Status 1 %",
    size="Average_Income",
    color="Neg_ammortization",
    title="Observed Default Rate by DTIR1 and Negative Amortization, with Income by Size",
    labels={
        "DTIR1": "DTIR1",
        "Status 1 %": "Default Rate (%)",
        "Average_Income": "Average Income",
        "Neg_ammortization": "Negative Amortization"
    },
    opacity=0.75,
    width=900,
    height=500
)

fig.update_xaxes(
    range=[0, 100],
    title="DTIR1"
)

fig.update_yaxes(
    range=[0, 100],
    ticksuffix="%"
)

fig.show()

<span style="color:cadetblue;"><b></b>
The observed default rate is consistently higher for **neg_amm** than for **not_neg**. Default rates are higher at DTI levels of 0–20, lower between 20–50, and increase again around 50–60, with some observations reaching 70% or higher. This suggests that negative amortization and DTI jointly show different patterns of observed default rates.</span>


### Calculate the observed default rate for each DTI interval.

In [152]:
# Create a separate EDA dataset with DTI ranges.
# The original df remains unchanged for modeling.

df_eda = df.copy()

# Calculate the observed default rate for each DTI interval.
# Borrower count helps us interpret the reliability of each interval.

dtir1_bins = [4, 10, 20, 30, 40, 50, 61]
dtir1_labels = ["5–10", "10–20", "20–30", "30–40", "40–50", "50–61"]

df_eda["DTI Range"] = pd.cut(
    df_eda["dtir1"],
    bins=dtir1_bins,
    labels=dtir1_labels,
    include_lowest=True
)

dtir1_summary = (
    df_eda.groupby("DTI Range", observed=True)["Status"]
          .agg(
              Borrower_Count="count",
              Default_Count="sum",
              Default_Rate="mean"
          )
          .reset_index()
)

dtir1_summary["Status 1 %"] = (
    dtir1_summary["Default_Rate"] * 100
).round(2)

display(
    dtir1_summary[
        ["DTI Range", "Borrower_Count", "Default_Count", "Status 1 %"]
    ]
)

,DTI Range,Borrower_Count,Default_Count,Status 1 %
0,5–10,2316,636,27.46
1,10–20,5594,1257,22.47
2,20–30,19176,2284,11.91
3,30–40,40472,4916,12.15
4,40–50,44530,6895,15.48
5,50–61,8334,3556,42.67


In [ ]:
# Plot the observed default percentage across DTI ranges.
# This allows us to see whether default rates are lower in the middle
# and higher toward the lower or upper ends of the DTI range.

import plotly.express as px

fig = px.bar(
    dtir1_summary,
    x="DTI Range",
    y="Status 1 %",
    text="Status 1 %",
    title="Observed Default Rate by DTI Range",
    labels={
        "DTI Range": "DTI Range",
        "Status 1 %": "Default Rate (%)"
    }
)

fig.update_traces(
    texttemplate="%{text:.1f}%",
    textposition="outside"
)

fig.update_yaxes(
    range=[0, 100],
    ticksuffix="%"
)

fig.show()

In [ ]:
# Compare observed default rates across loan amount ranges and loan types.
# X-axis represents average borrower income, Y-axis represents default rate,
# and color represents loan type.

import pandas as pd
import plotly.express as px

plot_data = df[
    ["loan_amount", "income", "loan_type", "Status"]
].dropna().copy()

# Divide loan amounts into 300 ranges.
plot_data["Loan Amount Range"] = pd.qcut(
    plot_data["loan_amount"],
    q=300,
    duplicates="drop"
)

# Calculate default rate and average income for each loan amount range and loan type.
summary = (
    plot_data
    .groupby(["Loan Amount Range", "loan_type"], observed=True)
    .agg(
        Default_Rate=("Status", "mean"),
        Average_Income=("income", "mean")
    )
    .reset_index()
)

summary["Status 1 %"] = summary["Default_Rate"] * 100

fig = px.scatter(
    summary,
    x="Average_Income",
    y="Status 1 %",
    color="loan_type",
    title="Observed Default Rate by Average Income and Loan Type",
    labels={
        "Average_Income": "Average Income",
        "Status 1 %": "Default Rate (%)",
        "loan_type": "Loan Type"
    },
    opacity=0.75,
    width=900,
    height=500
)

fig.update_yaxes(
    range=[0, 100],
    ticksuffix="%"
)

fig.show()

In [ ]:
# Compare observed default rates across loan amount ranges and lump-sum payment status.
# X-axis represents average borrower income, Y-axis represents default rate,
# and color represents lump-sum payment status.

import pandas as pd
import plotly.express as px

plot_data = df[
    ["loan_amount", "income", "lump_sum_payment", "Status"]
].dropna().copy()

# Divide loan amounts into 300 ranges.
plot_data["Loan Amount Range"] = pd.qcut(
    plot_data["loan_amount"],
    q=300,
    duplicates="drop"
)

# Calculate default rate and average income for each loan amount range and category.
summary = (
    plot_data
    .groupby(["Loan Amount Range", "lump_sum_payment"], observed=True)
    .agg(
        Default_Rate=("Status", "mean"),
        Average_Income=("income", "mean")
    )
    .reset_index()
)

summary["Status 1 %"] = summary["Default_Rate"] * 100

fig = px.scatter(
    summary,
    x="Average_Income",
    y="Status 1 %",
    color="lump_sum_payment",
    title="Observed Default Rate by Average Income and Lump-Sum Payment Status",
    labels={
        "Average_Income": "Average Income",
        "Status 1 %": "Default Rate (%)",
        "lump_sum_payment": "Lump-Sum Payment"
    },
    opacity=0.75,
    width=900,
    height=500
)

fig.update_yaxes(
    range=[0, 100],
    ticksuffix="%"
)

fig.show()

In [ ]:
# Compare observed default rates across loan amount ranges and negative amortization status.
# X-axis represents average borrower income, Y-axis represents default rate,
# and color represents negative amortization status.

plot_data = df[
    ["loan_amount", "income", "Neg_ammortization", "Status"]
].dropna().copy()

# Divide loan amounts into 300 ranges.
plot_data["Loan Amount Range"] = pd.qcut(
    plot_data["loan_amount"],
    q=300,
    duplicates="drop"
)

# Calculate default rate and average income for each loan amount range and category.
summary = (
    plot_data
    .groupby(["Loan Amount Range", "Neg_ammortization"], observed=True)
    .agg(
        Default_Rate=("Status", "mean"),
        Average_Income=("income", "mean")
    )
    .reset_index()
)

summary["Status 1 %"] = summary["Default_Rate"] * 100

fig = px.scatter(
    summary,
    x="Average_Income",
    y="Status 1 %",
    color="Neg_ammortization",
    title="Observed Default Rate by Average Income and Negative Amortization",
    labels={
        "Average_Income": "Average Income",
        "Status 1 %": "Default Rate (%)",
        "Neg_ammortization": "Negative Amortization"
    },
    opacity=0.75,
    width=900,
    height=500
)

fig.update_yaxes(
    range=[0, 100],
    ticksuffix="%"
)

fig.show()

<span style="color:cadetblue;"><b></b>
Income shows a generally negative association with observed default rates across most categories, but the strength and consistency of the pattern differ by category. The lpsm category is the main exception, showing high and widely distributed default rates without a clear income-related downward pattern.</span>

## 6. Preparing the Data for Modeling

After completing the cleaningb and exploratory analysis, the dataset will be prepared for machine learning.

In [ ]:
# Encode categorical features into 0/1 columns.

categorical_features = df.select_dtypes(include="object").columns.tolist()

df_encoded = pd.get_dummies(df,columns=categorical_features, drop_first=False)

# Convert True/False dummy values to 0/1.
dummy_columns = df_encoded.select_dtypes(include="bool").columns

df_encoded[dummy_columns] = df_encoded[dummy_columns].astype(int)

# Display the encoded data.
display(df_encoded.head())

In [158]:
# Separate predictor features from the target and split the data.

from sklearn.model_selection import train_test_split

X = df_encoded.drop(columns=["Status"])
y = df_encoded["Status"]

X_train, X_test, y_train, y_test = train_test_split(X, y,test_size=0.20,random_state=42,stratify=y)

print("Training set:", X_train.shape)
print("Testing set:", X_test.shape)

Training set: (96337, 60)
Testing set: (24085, 60)


In [ ]:
# Scale the numerical features.

from sklearn.preprocessing import StandardScaler

numerical_features = df.select_dtypes(
    include=["int64", "float64"]
).columns.tolist()

numerical_features.remove("Status")

scaler = StandardScaler()

X_train[numerical_features] = scaler.fit_transform(X_train[numerical_features])

X_test[numerical_features] = scaler.transform(X_test[numerical_features])

# Display the prepared training data.
display(X_train.head())

## 7. Baseline Classification Model

A baseline classification model will be developed to provide an initial benchmark for predicting loan default. The purpose of this model is to provide a measurable starting point against which additional models, improvements, and optimization techniques can be compared.
For Model evaluation and performance metric, the classification model will be evaluated using accuracy, precision, recall, and F1 score. Since the data is imbalanced, multiple metrics are used instead of relying on accuracy alone. A confusion matrix will also be used to examine correct and incorrect predictions and support comparison with other models and future improvements.

In [ ]:
# Build and evaluate the Logistic Regression model.

import time

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    confusion_matrix
)

logistic_model = LogisticRegression(
    class_weight="balanced",
    max_iter=1000
)

# Measure training time.
start_train = time.perf_counter()

logistic_model.fit(X_train, y_train)

training_time = time.perf_counter() - start_train

# Measure testing/prediction time.
start_test = time.perf_counter()

y_pred = logistic_model.predict(X_test)

testing_time = time.perf_counter() - start_test

# Calculate evaluation metrics.
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, zero_division=0)
recall = recall_score(y_test, y_pred, zero_division=0)
f1 = f1_score(y_test, y_pred, zero_division=0)

# Store results.
logistic_results = pd.DataFrame({
    "Model": ["Logistic Regression"],
    "Accuracy": [accuracy],
    "Precision": [precision],
    "Recall": [recall],
    "F1 Score": [f1],
    "Training Time (seconds)": [training_time],
    "Testing Time (seconds)": [testing_time]
})

display(logistic_results.round(4))

In [ ]:
# Visualize the Logistic Regression confusion matrix.

import matplotlib.pyplot as plt
import seaborn as sns

cm = confusion_matrix(y_test, y_pred)

plt.figure(figsize=(5, 3))

sns.heatmap(
    cm,
    annot=True,
    fmt="d",
    cmap="Blues",
    xticklabels=["No Default (0)", "Default (1)"],
    yticklabels=["No Default (0)", "Default (1)"]
)

plt.title("Logistic Regression Confusion Matrix")
plt.xlabel("Predicted Status")
plt.ylabel("Actual Status")
plt.tight_layout()
plt.show()